File: backend/src/main/java/com/skewwhiffy/auraltester/notes/Interval.java
package com.skewwhiffy.auraltester.notes;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.val;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;

@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class Interval {
  private static final List<String> displayStrings = Arrays.asList(
    "unison",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "octave"
  );

  private static final Collection<Integer> perfectDegrees = Arrays.asList(1, 4, 5, 8);

  private final int degree;
  public int getDegree() {
    return degree;
  }
  private final int deviation;
  public int getDeviation() {
    return deviation;
  }

  Interval getDiminished() {
    return new Interval(degree, deviation - 1);
  }

  Interval getAugmented() {
    return new Interval(degree, deviation + 1);
  }

  String getDisplayString() {
    return String.format("%s %s", getQuality(), displayStrings.get(degree - 1));
  }

  private String getQuality() {
    if (deviation == 0) {
      return perfectDegrees.contains(degree) ? "perfect" : "major";
    }
    if (deviation == - 1 && !perfectDegrees.contains(degree)) {
      return "minor";
    }
    if (deviation < 0) {
      val diminishedDegree = perfectDegrees.contains(degree) ? -deviation : -deviation - 1;
      if (diminishedDegree == 1) {
        return "diminished";
      }
      if (diminishedDegree == 2) {
        return "doubly diminished";
      }
      return diminishedDegree + "x diminished";
    }
    if (deviation == 1) {
      return "augmented";
    }
    if (deviation == 2) {
      return "doubly augmented";
    }
    return deviation + "x augmented";
  }

  static Interval augmented(int degree) {
    val baseInterval = perfectDegrees.contains(degree) ? perfect(degree) : major(degree);
    return baseInterval.getAugmented();
  }

  static Interval diminished(int degree) {
    val baseInterval = perfectDegrees.contains(degree) ? perfect(degree) : minor(degree);
    return baseInterval.getDiminished();
  }

  static Interval minor(int degree) {
    return major(degree).getDiminished();
  }

  static Interval major(int degree) {
    if (perfectDegrees.contains(degree)) {
      throw new IllegalArgumentException("Cannot instantiate major interval of degree " + degree);
    }
    return new Interval(degree, 0);
  }

  static Interval perfect(int degree) {
    if (!perfectDegrees.contains(degree)) {
      throw new IllegalArgumentException("Cannot instantiate perfect interval of degree " + degree);
    }
    return new Interval(degree, 0);
  }
}

File: backend/src/main/scala/com/skewwhiffy/auraltester/notes/Interval.scala
package com.skewwhiffy.auraltester.notes

object Interval {
  lazy val displayStrings: List[String] = List(
    "unison",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "octave"
  )

  lazy val perfectDegrees: Set[Int] = Set(1, 4, 5, 8)

  lazy val augmented: Int => Interval = degree => {
    val baseInterval = if (perfectDegrees.contains(degree)) perfect(degree) else major(degree)
    baseInterval.augmented
  }

  lazy val diminished: Int => Interval = degree => {
    val baseInterval = if (perfectDegrees.contains(degree)) perfect(degree) else minor(degree)
    baseInterval.diminished
  }

  lazy val minor: Int => Interval = degree => major(degree).diminished

  lazy val major: Int => Interval = degree => {
    if (perfectDegrees.contains(degree)) {
      throw new IllegalArgumentException(s"Cannot instantiate major interval of degree '$degree'")
    }
    new Interval(degree, 0)
  }

  lazy val perfect: Int => Interval = degree => {
    if (!perfectDegrees.contains(degree)) throw new IllegalArgumentException(s"Cannot instantiate perfect interval of degree '$degree'")
    else new Interval(degree, 0)
  }

  class Interval(val degree: Int, val deviation: Int) {
    lazy val diminished: Interval = new Interval(degree, deviation - 1)

    lazy val augmented: Interval = new Interval(degree, deviation + 1)

    lazy val displayString: String = s"$quality ${displayStrings(degree - 1)}"

    lazy val up: DirectedInterval = new DirectedInterval(this, IntervalDirection.Up)

    lazy val down: DirectedInterval = new DirectedInterval(this, IntervalDirection.Down)

    private def quality: String = {
      lazy val defaultQuality = if (perfectDegrees.contains(degree)) "perfect" else "major"
      lazy val negativeQuality = {
        if (!perfectDegrees.contains(degree) && deviation == -1) "minor"
        else {
          (if (perfectDegrees.contains(degree)) -deviation else -deviation - 1) match {
            case 1 => "diminished"
            case 2 => "doubly diminished"
            case it => s"${it}x diminished"
          }
        }
      }
      lazy val positiveQuality = (deviation match {
        case 1 => ""
        case 2 => "doubly "
        case it => s"${it}x "
      }) + "augmented"

      deviation match {
        case 0 => defaultQuality
        case it if it < 0 => negativeQuality
        case it if it > 0 => positiveQuality
      }
    }

    override def equals(obj: Any): Boolean = obj match {
      case other: Interval => other.deviation == deviation && other.degree == degree
      case _ => false
    }

    override def toString: String = displayString
  }
}